README.md
# Proxmox Inventory Tool

This tool generates a list of vms in one or more proxmox clusters and consolidates the information into a JSON format.

## Features

### Proxmox API

Calls the Proxmox API to generate a List of VM's and their IP addresses in a given set of clusters.

At the moment can only retreive IP's of virtual machines that support cloud init.

### HTTP endpoint

The Tool Provides some http endpoints to retreive the generated JSON

| Path | HTTP Method | Form Parameters | Response |
|--|--|--|--|
|`/`| `GET` | - | Catch all returning `NotImplemented` |
|`/all`| `GET` | - | JSON containing all clusters and all VMs |
|`/cluster`| `POST` | `cluster=<CLUSTER_NAME>` | JSON containing VMs of a specific cluster |

#### Examples

```sh
# Get the whole list
curl "localhost:8080/all"

# Get the cluster specific list
curl "localhost:8080/cluster" --form-string 'cluster=MY_CLUSTER'
```

## Config

You'll have to create an config file as the `example.yaml` and fill out the values as needed.

The config file has to contain a `clusters` list in YAML format. All other options can be passed as ENV vars or as CLI Flag.

```yaml
# -- Cluster list to fetch info from
clusters:
  - # Name must be unique
    name: "CUSTOMN CLUSTER NAME"
    # API key for the proxmox cluster
    apikey: "aaaaaaaaa-bbb-cccc-dddd-ef0123456789"
    # API user
    apiuser: "root@pam!monitoring"
    # API URL
    apihost: "https://10.0.0.1:8006"
```

### CLI Flags

You can also pass most of the values as CLI Flag or Environment Variable.

Note that the Config File always has to be specified.

|Flag | Long Flag | Default Value | ENV_VAR | Usage
|--|--|--|--|--|
|-f|--config-file|`-`|INV_CONFIG_FILE|YAML file containing your config values. Values set here override all commandline flags and environment vars|
|-l|--listen-address|`:8080`|INV_HTTP_LISTEN|The http port to listen on|
|-L|--log-level|`INFO`|INV_LOG_LEVEL|Set the Log Level / verbosity|
|-t|--timeout|`10`|INV_TIMEOUT|Time in seconds before a request times out|
|-i|--fetch-interval|`300`|INV_INTERVAL|Interval at whicht to refetch all VMs|

## Build Instructions

To build the binary yourself you'll have to have `go 1.X` installed.

```sh
CGO_ENABLED=0 go mod download
CGO_ENABLED=0 go build -o pveinventory cmd/main.go
```

CHANGELOG.md
## [1.2.1](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/compare/v1.2.0...v1.2.1) (2023-07-31)


### Bug Fixes

* logging caller report ([00d8b15](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/00d8b150726a4e1646e3b508a464ddcea52416c2))
* logging caller report ([ba7a059](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/ba7a059e21c3a443a1c2d63d0f07af0f5a479ee6))
* typo in json struct ([389b6b8](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/389b6b869e6632f2332a9858deb0b2dc9fb72531))
* typo in json struct ([7cf540a](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/7cf540a7a5e4e0029dc1012ed2ff188193d411ed))

# [1.2.0](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/compare/v1.1.1...v1.2.0) (2023-07-21)


### Features

* **web:** add collapsible item list ([76910b8](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/76910b8461d6f211036ed898150455be79f4c00d))
* **web:** add sortable table ([66ad4bc](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/66ad4bc2bb066d9457eac9ca4bf9466701a88094))

# [1.2.0-rc.2](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/compare/v1.2.0-rc.1...v1.2.0-rc.2) (2023-07-21)


### Features

* **web:** add collapsible item list ([85fe158](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/85fe158449fab4ddc208c0613c722dcc9fd99130))

# [1.2.0-rc.1](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/compare/v1.1.1-rc.1...v1.2.0-rc.1) (2023-07-20)


### Features

* **web:** add sortable table ([c4a5550](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/c4a55505d61f314724296a3002f6a6647e0e6a0a))

## [1.1.1](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/compare/v1.1.0...v1.1.1) (2023-07-14)


### Bug Fixes

* logging and ci stuff ([0dcae31](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/0dcae319c635809e22df7f97ff0c454332e90218))
* logging and ci stuff ([52c27d6](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/52c27d6ef934e53be8cd8768e6d643f6d7d484c1))

## [1.1.1-rc.1](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/compare/v1.1.0...v1.1.1-rc.1) (2023-07-14)


### Bug Fixes

* logging and ci stuff ([0dcae31](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/0dcae319c635809e22df7f97ff0c454332e90218))

# [1.1.0](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/compare/v1.0.1...v1.1.0) (2023-07-14)


### Bug Fixes

* **web:** commented out unused code ([0504e59](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/0504e59cf45e20e9927e62566d6f853635484a67))
* **web:** log,url,html formatting ([2fd94d4](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/2fd94d498faed06482fce2bfd14a487c3d0c2b3b))


### Features

* **web:** add a basic web interface ([295ff2c](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/295ff2c6189aeac93e12a5440efb793d5770ab70))

# [1.1.0-rc.2](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/compare/v1.1.0-rc.1...v1.1.0-rc.2) (2023-07-14)


### Bug Fixes

* **web:** log,url,html formatting ([2fd94d4](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/2fd94d498faed06482fce2bfd14a487c3d0c2b3b))

# [1.1.0-rc.1](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/compare/v1.0.1...v1.1.0-rc.1) (2023-07-13)


### Bug Fixes

* **web:** commented out unused code ([0504e59](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/0504e59cf45e20e9927e62566d6f853635484a67))


### Features

* **web:** add a basic web interface ([295ff2c](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/295ff2c6189aeac93e12a5440efb793d5770ab70))

## [1.0.1](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/compare/v1.0.0...v1.0.1) (2023-06-08)


### Bug Fixes

* ci ([f5308f5](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/f5308f5155a56ae48da4d36b0e60f6c384c4932e))
* ci ([8ce5bb4](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/8ce5bb4e1e8045d1ea7974c6314e690748a1ac75))
* ci ([4fc5cb8](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/4fc5cb8718853205aa32d5988c9a1367a7c533a5))
* ci ([f900cba](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/f900cba208991a91682b0bd2a65c8d38949f7f4d))
* ci ([e8bbae1](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/e8bbae1f0900bb9a8da4e1849f14af3bbea98ffa))
* ci ([a2de26d](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/a2de26dccce773595b8013566f03ea76b4fab68a))
* config initialization ([84d0a59](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/84d0a591ce22df8a80b3d8ec070d14f87d9128ec))
* config initialization ([7d09395](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/7d09395d7284b7b68da0ce4b60b8e5ec8d098c33))

## [1.0.1-dev.4](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/compare/v1.0.1-dev.3...v1.0.1-dev.4) (2023-06-08)


### Bug Fixes

* ci ([f5308f5](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/f5308f5155a56ae48da4d36b0e60f6c384c4932e))
* ci ([8ce5bb4](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/8ce5bb4e1e8045d1ea7974c6314e690748a1ac75))
* ci ([4fc5cb8](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/4fc5cb8718853205aa32d5988c9a1367a7c533a5))

## [1.0.1-dev.3](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/compare/v1.0.1-dev.2...v1.0.1-dev.3) (2023-06-08)


### Bug Fixes

* ci ([f900cba](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/f900cba208991a91682b0bd2a65c8d38949f7f4d))
* ci ([e8bbae1](https://gitlab.cloudical.net/operations/proxmox-vm-inventory/commit/e8bbae1f0900bb9a8da4e1849f14af3bbea98ffa))

pkg/config/config.go
package config

import (
	"io"
	"os"

	"github.com/alecthomas/kingpin/v2"
	"github.com/charmbracelet/log"
	"github.com/mitchellh/mapstructure"
	"gopkg.in/yaml.v3"
)

type Config struct {
	RequestTimeout    *int       `mapstructure:"timeout"`
	FetchInterval     *int       `mapstructure:"fetchInterval"`
	HttpListenAddress *string    `mapstructure:"httpAddress"`
	LogLevel          *string    `mapstructure:"logLevel"`
	Clusters          *[]Cluster `mapstructure:"clusters"`
}

type Cluster struct {
	Name    string `mapstructure:"name" json:"name"`
	ApiKey  string `mapstructure:"apikey" json:"apikey,omitempty"`
	ApiUser string `mapstructure:"apiuser" json:"apiuser,omitempty"`
	ApiHost string `mapstructure:"apihost" json:"apihost"`
}

// variables to be parsed
var filePath *string = kingpin.
	Flag("config-file", "YAML file containing your config values. Values set here override all command line flags and environment vars").
	Short('f').
	Required().
	Envar("INV_CONFIG_FILE").
	String()

var c *Config = &Config{
	FetchInterval: kingpin.
		Flag("fetch-interval", "Interval at which to refetch all VMs").
		Short('i').
		Default("300").
		Envar("INV_INTERVAL").
		Int(),
	RequestTimeout: kingpin.
		Flag("timeout", "Time in seconds before a request times out").
		Short('t').
		Default("10").
		Envar("INV_TIMEOUT").
		Int(),
	HttpListenAddress: kingpin.
		Flag("listen-address", "The http port to listen on").
		HintOptions(":8080", "127.0.0.1:8080", "[::]:8080").
		Default(":8080").
		Short('l').
		Envar("INV_HTTP_LISTEN").
		String(),
	LogLevel: kingpin.
		Flag("log-level", "Set the Log Level / verbosity").
		Short('L').
		Envar("INV_LOG_LEVEL").
		Default("INFO").
		Enum("DEBUG", "INFO", "WARN", "ERROR", "FATAL"),
}

// Creates a new Config file
// reads commandline flags and parses them
// any values in the provided yaml overwrite CLI flags
func New() *Config {

	kingpin.Parse()

	// setting log level for library
	level := log.ParseLevel(*c.LogLevel)
	log.SetReportCaller(true)
	log.SetLevel(level)
	log.Info("Setting Log Level", "level", *c.LogLevel)

	f, err := os.Open(*filePath)
	defer func(f *os.File) {
		if err := f.Close(); err != nil {
			log.Error("closing file failed", "err", err)
		}
	}(f)

	if err != nil {
		log.Fatal("could not open config file", "err", err)
	}

	r, err := io.ReadAll(f)
	if err != nil {
		log.Fatal("cold not read content", "err", err)
	}

	// unmarshal yaml into interface
	var y interface{}
	if err := yaml.Unmarshal(r, &y); err != nil {
		log.Fatal("cold not read YAML", "err", err)
	}

	// decoding yaml into struct
	decoder, _ := mapstructure.NewDecoder(&mapstructure.DecoderConfig{WeaklyTypedInput: true, Result: &c})
	if err := decoder.Decode(y); err != nil {
		log.Fatal("error unmarshalling YAML", "err", err)
	}

	return c
}

func ClusterList() []Cluster {
	l := make([]Cluster, len(*c.Clusters))

	for i, v := range *c.Clusters {
		l[i] = Cluster{
			Name:    v.Name,
			ApiHost: v.ApiHost,
		}
	}

	return l
}
